### PR TITLE
Improved Encoding Handling for Jobs

### PR DESCRIPTION
**Pull request by jmchilton**
Fixes some bugs at least with sqlite. New tool that demonstrates this stuff and that has tests which fail without these changes:

```
./run_tests -framework -id unicode_stream
```

### CODE REVIEW

**Comment by jmchilton** [comment]
Ping @bgruening - can you try the failing planemo khmer wrappers against this branch of Galaxy. Just add my remote, 

```
git remote add jmchilton https://github.com/jmchilton/galaxy.git
git checkout unicode_streams_3
cd path/to/khmer/
planemo t --galaxy_root path/to/galaxy
```

Should fix https://github.com/galaxyproject/galaxy/issues/714 I suppose. Ping @mr-c.

File: lib/galaxy/jobs/__init__.py [hunk 1]
```diff
@@ -869,7 +869,7 @@ def get_special( ):
         self.dependency_shell_commands = self.tool.build_dependency_shell_commands()
         # We need command_line persisted to the db in order for Galaxy to re-queue the job
         # if the server was stopped and restarted before the job finished
-        job.command_line = self.command_line
+        job.command_line = util.unicodify(self.command_line)
```
Comment (nsoranzo) [inline lib/galaxy/jobs/__init__.py]: `unicodify` is already explicitly imported (it's used by the `finish()` method). It'd be nice to consistently use one or the other.

**Comment by jmchilton** [comment]
@nsoranzo Updated for your comments and fixed some failing unit tests.

**Comment by nsoranzo** [comment]
You can remove the following lines from `lib/galaxy/jobs/__init__.py`:

```
DATABASE_MAX_STRING_SIZE = util.DATABASE_MAX_STRING_SIZE
DATABASE_MAX_STRING_SIZE_PRETTY = util.DATABASE_MAX_STRING_SIZE_PRETTY
```

**Comment by jmchilton** [comment]
@nsoranzo Done, thanks or the catch.

File: lib/galaxy/jobs/__init__.py [hunk 2]
```diff
@@ -1839,13 +1822,7 @@ def finish( self, stdout, stderr, tool_exit_code=None ):
             task.state = task.states.ERROR
 
         # Save stdout and stderr
-        if len( stdout ) > DATABASE_MAX_STRING_SIZE:
-            log.error( "stdout for task %d is greater than %s, only a portion will be logged to database" % ( task.id, DATABASE_MAX_STRING_SIZE_PRETTY ) )
```
Comment (nsoranzo) [inline lib/galaxy/jobs/__init__.py:1843]: Why was `log.error()` used here? :open_mouth: 
Anyway, now it's rightfully replaced with `log.info()`.

File: lib/galaxy/jobs/__init__.py [hunk 3]
```diff
@@ -1839,13 +1822,7 @@ def finish( self, stdout, stderr, tool_exit_code=None ):
...
-        task.stdout = util.shrink_string_by_size( stdout, DATABASE_MAX_STRING_SIZE, join_by="\n..\n", left_larger=True, beginning_on_size_error=True )
-        if len( stderr ) > DATABASE_MAX_STRING_SIZE:
-            log.error( "stderr for task %d is greater than %s, only a portion will be logged to database" % ( task.id, DATABASE_MAX_STRING_SIZE_PRETTY ) )
-        self._collect_metrics( task )
```
Comment (nsoranzo) [inline lib/galaxy/jobs/__init__.py:1847]: Here you are removing also a call to `_collect_metrics()` method, is this correct?

File: lib/galaxy/model/__init__.py [hunk 4]
```diff
@@ -130,6 +130,18 @@ def metrics( self ):
         # TODO: Make iterable, concatenate with chain
         return self.text_metrics + self.numeric_metrics
 
+    def set_streams( self, stdout, stderr ):
+        stdout = galaxy.util.unicodify( stdout )
+        stderr = galaxy.util.unicodify( stderr )
+        if ( len( stdout ) > galaxy.util.DATABASE_MAX_STRING_SIZE ):
+            stdout = galaxy.util.shrink_string_by_size( stdout, galaxy.util.DATABASE_MAX_STRING_SIZE, join_by="\n..\n", left_larger=True, beginning_on_size_error=True )
+            log.info( "stdout for %s %d is greater than %s, only a portion will be logged to database" % ( type(self), self.id, galaxy.util.DATABASE_MAX_STRING_SIZE_PRETTY ) )
```
Comment (blankenberg) [inline lib/galaxy/model/__init__.py]: If moving around/changing these log commands, we should also stop doing the in-place string substitution, and let the log method handle it.

File: lib/galaxy/jobs/output_checker.py [hunk 5]
```diff
@@ -134,8 +134,7 @@ def check_output( tool, stdout, stderr, tool_exit_code, job ):
 
     # Store the modified stdout and stderr in the job:
     if None is not job:
```
Comment (nsoranzo) [inline lib/galaxy/jobs/output_checker.py]: `if job is not None:` sounds better.

**Comment by mr-c** [comment]
@jmchilton Yay! Thank you.

**Comment by bgruening** [comment]
@jmchilton does not fix it for the khmer case. Let's talk about this tomorrow.

**Comment by bgruening** [comment]
But the good news is that this test tool is failing for me. So we are pretty close to figure out what is happening.

**Comment by jmchilton** [comment]
@blankenberg and @nsoranzo begrudgingly "fixed" your complaints in  8aab57a. Thanks for the thorough review though.

**Comment by nsoranzo** [comment]
Thanks @jmchilton! Let's wait for @bgruening's approval.

**Comment by bgruening** [comment]
:+1:

**Comment by bgruening** [comment]
Thanks a lot!

**Comment by jmchilton** [comment]
So this does fix #714 I think, but it doesn't resolve the planemo issue. This is fine though, so @bgruening  and I decided to get this merged so that the planemo issue isn't confounded with these issues.

**Pull request merged**